src/libcore/iter.rs

```diff
@@ -369,6 +369,21 @@ pub trait Iterator {
     /// `None` here means that either there is no known upper bound, or the
     /// upper bound is larger than `usize`.
     ///
+    /// # Implementation notes
+    ///
+    /// It is not enforced that an iterator implementation yields
+    /// declared number of elements. A buggy iterator may yield less
+    /// than lower bound or more than upper bound of elements.
+    ///
+    /// `size_hint()` can be used to optimize the fast path, and can be
+    /// used to panic if estimation is incorrect, but must not be used,
+    /// e. g. to omit bound checks in unsafe operations. An incorrect
```

Comment:
This is a bit in the weeds, but I'd typically write this as "...must not be used to e.g. omit bound checks..."